### PR TITLE
PTV-1887 - linting, formatting, updating reps, fixing tests

### PR DESCRIPTION
**Pull request by eapearson**
This set of changes aims to put the codebase into a stable, updated state in order to be able to make further changes.
The heart of, and initial set of, the changes was to apply uniform formatting via black, linting via pylint and sonarlint. 
Once these fixes were made, tests were run and any broken tests fixed until there were no errors.
However, once tests were passing, it was clear that not all tests were being invoked - code coverage was at about 50% and there were many tests skipped due to incorrect async usage and a few other issues. In fact, code coverage reports were not being generated either.
So the next effort was to update all dependencies, including adding a missing dependency required for asynchronous testing. Once through this, coverage was up to 84%.
Finally, once the changes were captured in a pull request, a series of changes were made to address any issues raised by sonar cloud or codify until they were all resolved. Some of these issues are specific to repeated values used in tests, which were simply annotated with # NOSONAR rather than rework the tests, as there are dozens if not hundreds of inline test data, and probably a 30-40 instances of duplicated test data which needed annotating.
There are no functional changes made around design. All code changes were made to address code quality issues raised by linting and related tools.

I did not include some changes that would help keep the codebase in shape. E.g. there should be tools (preferably docker-based) to apply formatting, linting checks, certainly ad-hoc but also via pre-commit hook and in the GitHub actions. Other welcome additions would be tightening up the type annotations, including mypy (currently 69 errors, and would require several more dependencies), and separating developer dependencies from those required by the service (runtime.)

### CODE REVIEW

**Comment by eapearson** [comment]
Hello @bio-boris. Here are the changes without the functional changes to fix and improve uploading. I'll start working on those next,

**Comment by eapearson** [comment]
@bio-boris I've made a few changes since yesterday. Mostly just around release notes, and a few odds and ends.
It would be nice to get control over codacy and other code quality tools to get them to align with developer tools. It mostly does, but e.g. codacy uses "remark-lint" whereas markdownlint seems more popular and widely supported.

After this, I'll have a PR with the functional changes for fixing upload.

File: import_specifications/generate_import_template.py [hunk 1]
```diff
@@ -20,45 +20,64 @@
 _NMS_URLS = {
```
Comment (bio-boris) [inline import_specifications/generate_import_template.py:20]: LGTM

File: requirements.txt [hunk 2]
```diff
@@ -1,17 +1,20 @@
...
-xlrd==2.0.1
-openpyxl==3.0.7
+aiohttp==3.8.4
+black==23.7.0
+coverage==7.2.7
 defusedxml==0.7.1
-python-magic==0.4.24
\ No newline at end of file
+frozendict==2.3.8
+globus-sdk==3.23.0
```
Comment (bio-boris) [inline requirements.txt]: hopefully this globus update still 
works

File: run_tests.sh [hunk 3]
```diff
@@ -2,4 +2,13 @@
 DIR="$( cd "$( dirname "$0" )" && pwd )"
```
Comment (bio-boris) [inline run_tests.sh:2]: LGTM

File: scripts/prune_acls.py [hunk 4]
```diff
@@ -117,13 +127,13 @@ def get_old_acls():
...
+        directory = "/dtn/disk0/bulk" + acl["path"]
+        if directory_is_old(directory) and acl["id"] not in admin_acls:
             oad = old_acl_and_dir(acl, directory)
             old_acls.append(oad)
 
     return old_acls
 
 
-if __name__ == '__main__':
+if __name__ == "__main__":
```
Comment (bio-boris) [inline scripts/prune_acls.py:138]: LGTM

File: scripts/refresh_token.py [hunk 5]
```diff
@@ -36,52 +36,54 @@
...
 
 print((authorizer.access_token))
-parser.set('general', 'auth_token', authorizer.access_token)
+parser.set("general", "auth_token", authorizer.access_token)
 
 # Writing our configuration file
-#with open('/root/.globus.cfg', 'wb') as configfile:
-with open('/root/.globus.cfg', 'w') as configfile:
+# with open('/root/.globus.cfg', 'wb') as configfile:
+with open("/root/.globus.cfg", "w") as configfile:
```
Comment (bio-boris) [inline scripts/refresh_token.py:88]: LGTM

File: staging_service/AutoDetectUtils.py [hunk 6]
```diff
@@ -1,15 +1,44 @@
...
+
+# pylint: disable=C0115,C0116
 
 
 class AutoDetectUtils:
-    _MAPPINGS = None  # expects to be set by config
+    _MAPPINGS: Union[None, dict[str, Any]] = None  # expects to be set by config
+
+    @classmethod
+    def set_mappings(cls, mappings_config: MappingsConfig):
```
Comment (bio-boris) [inline staging_service/AutoDetectUtils.py:19]: I'm a bit confused about why these are being changed to class methods, as the rest of the codebase uses static methods.

File: staging_service/JGIMetadata.py [hunk 7]
```diff
@@ -1,22 +1,20 @@
-from .utils import Path
+import os
```
Comment (bio-boris) [inline staging_service/JGIMetadata.py:1]: LGTM

File: staging_service/__main__.py [hunk 8]
```diff
@@ -1,9 +1,11 @@
-from .app import app_factory
-from aiohttp import web
 import asyncio
```
Comment (bio-boris) [inline staging_service/__main__.py:1]: LGTM

File: staging_service/app_error_formatter.py [hunk 9]
```diff
@@ -4,7 +4,7 @@
 
 from pathlib import Path
 
-from .import_specifications.file_parser import ErrorType, Error
+from .import_specifications.file_parser import Error, ErrorType
```
Comment (bio-boris) [inline staging_service/app_error_formatter.py:7]: LGTM

File: staging_service/auth2Client.py [hunk 10]
```diff
@@ -4,13 +4,14 @@
 @author: gaprice@lbl.gov
 modified for python3 and authV2
```
Comment (bio-boris) [inline staging_service/auth2Client.py:5]: LGTM

File: staging_service/autodetect/GenerateMappings.py [hunk 11]
```diff
@@ -20,85 +21,111 @@
...
 
 * Note: We should serve the generated content from memory
-* Note: This doesn't handle if we want to have different output types based on file extensions feeding into the same app
+* Note: This doesn't handle if we want to have different output types based on file extensions
+  feeding into the same app
 """
 from collections import defaultdict
 
-from staging_service.autodetect.Mappings import *
+from staging_service.autodetect import Mappings
```
Comment (bio-boris) [inline staging_service/autodetect/GenerateMappings.py]: LGTM

File: staging_service/autodetect/GenerateMappings.py [hunk 12]
```diff
@@ -20,85 +21,111 @@
...
-    sample_set_id: "Samples",
-    media_id: "Media",
-    expression_matrix_id: "Expression Matrix",
-    metabolic_annotations_id: "Metabolic Annotations",
-    metabolic_annotations_bulk_id: "Bulk Metabolic Annotations",
-    fba_model_id: "FBA Model",
-    phenotype_set_id: "Phenotype Set",
-    escher_map_id: "EscherMap",
-    import_specification: "Import Specification",
+    Mappings.sra_reads_id: "SRA Reads",
```
Comment (bio-boris) [inline staging_service/autodetect/GenerateMappings.py]: Sad that these are long now :(

Comment (bio-boris) [reply staging_service/autodetect/GenerateMappings.py]: These really make the readability much worse.

File: staging_service/autodetect/Mappings.py [hunk 13]
```diff
@@ -31,7 +31,8 @@
 
 MEDIA = "KBaseBiochem.Media"
 
-# These ID mappings should be available in dropdown_order in staging_upload.json in the narrative service
+# These ID mappings should be available in dropdown_order in staging_upload.json
```
Comment (bio-boris) [inline staging_service/autodetect/Mappings.py:34]: LGTM

File: staging_service/globus.py [hunk 14]
```diff
@@ -1,8 +1,10 @@
-from .utils import Path
-import aiohttp
-import aiofiles
-import os
 import configparser
```
Comment (bio-boris) [inline staging_service/globus.py:1]: Seems fine

File: staging_service/import_specifications/file_parser.py [hunk 15]
```diff
@@ -6,12 +6,10 @@
 from collections.abc import Callable
```
Comment (bio-boris) [inline staging_service/import_specifications/file_parser.py:6]: LGTM

File: staging_service/import_specifications/file_writers.py [hunk 16]
```diff
@@ -220,11 +248,14 @@ def _write_excel_row(sheet: Worksheet, row: int, contents: list[Any]):
...
 
+
 def _expand_excel_columns_to_max_width(sheet: Worksheet):
     # https://stackoverflow.com/a/40935194/643675
     for column_cells in sheet.columns:
         length = max(len(_as_text(cell.value)) for cell in column_cells)
-        sheet.column_dimensions[get_column_letter(column_cells[0].column)].width = length
+        sheet.column_dimensions[
+            get_column_letter(column_cells[0].column)
+        ].width = length
```
Comment (bio-boris) [inline staging_service/import_specifications/file_writers.py:258]: lgtm

File: staging_service/import_specifications/individual_parsers.py [hunk 17]
```diff
@@ -31,10 +30,14 @@
...
-_HEADER_REGEX = re.compile(f"{_DATA_TYPE} (\\w+){_HEADER_SEP} "
-    + f"{_COLUMN_STR} (\\d+){_HEADER_SEP} {_VERSION_STR} (\\d+)")
+_EXPECTED_HEADER = (
+    f"{_DATA_TYPE} <data_type>{_HEADER_SEP} "
+    + f"{_COLUMN_STR} <column count>{_HEADER_SEP} {_VERSION_STR} <version>"
+)
+_HEADER_REGEX = re.compile(
+    f"{_DATA_TYPE} (\\w+){_HEADER_SEP} "
+    + f"{_COLUMN_STR} (\\d+){_HEADER_SEP} {_VERSION_STR} (\\d+)"
+)
```
Comment (bio-boris) [inline staging_service/import_specifications/individual_parsers.py:41]: LGTM

File: staging_service/metadata.py [hunk 18]
```diff
@@ -28,18 +30,18 @@ async def stat_data(path: Path) -> dict:
 
 def _file_read_from_tail(file_path):
     upper_bound = min(1024, os.stat(file_path).st_size)
-    with open(file_path, "r") as file:
+    with open(file_path, "r", encoding="utf-8") as file:
```
Comment (bio-boris) [inline staging_service/metadata.py:33]: Hopefully the platform encoding  was UTF-8!

File: staging_service/metadata.py [hunk 19]
```diff
@@ -75,7 +77,7 @@ async def add_upa(path: Path, UPA: str):
             data = await extant.read()
             data = decoder.decode(data)
     else:
-        data = await _generate_metadata(path)  # TODO performance optimization
+        data = await _generate_metadata(path, None)
```
Comment (bio-boris) [inline staging_service/metadata.py:80]: How did this ever work if _generate_metadata expected 2 inputs?

File: staging_service/utils.py [hunk 20]
```diff
@@ -187,7 +189,7 @@ def _add_acl(self, user_identity_id: str, shared_directory_basename: str):
 
             logging.info(response)
             logging.info(
-                "Shared %s with %s\n" % (shared_directory_basename, user_identity_id)
+                "Shared %s with %s\n", shared_directory_basename, user_identity_id
```
Comment (bio-boris) [inline staging_service/utils.py:192]: could use fstrings here as well

File: requirements.txt [hunk 21]
```diff
@@ -1,17 +1,19 @@
...
-xlrd==2.0.1
-openpyxl==3.0.7
+aiohttp==3.8.4
+black==23.7.0
+coverage==7.2.7
 defusedxml==0.7.1
-python-magic==0.4.24
\ No newline at end of file
+frozendict==2.3.8
+globus-sdk==1.6.1
```
Comment (bio-boris) [inline requirements.txt:8]: LGTM

File: tests/import_specifications/test_file_parser.py [hunk 22]
```diff
@@ -242,25 +274,28 @@ def test_ParseResults_init_w_error_success():
     assert pr == ParseResults(errors=errors_copy)
 
 
-def test_ParseResults_init_fail():
-    err = 'Exectly one of results or errors must be supplied'
-    parseResults_init_fail(None, None, ValueError(err))
-    parseResults_init_fail(PR_RESULTS, PR_ERROR, ValueError(err))
+def test_parse_results_init_fail():
```
Comment (bio-boris) [inline tests/import_specifications/test_file_parser.py:277]: LGTM

File: tests/import_specifications/test_file_parser.py [hunk 23]
```diff
@@ -406,50 +468,72 @@ def test_parse_import_specification_multiple_specs_and_parser_error():
...
+            call(Path("somefile.csv")),
+            call(Path("x.tsv")),  # NOSONAR python:S1192
+        ]
+    )
     parser1.assert_called_once_with(Path("myfile.xlsx"))
     parser2.assert_called_once_with(Path("somefile.csv"))
-    parser3.assert_called_once_with(Path("x.tsv"))
-    logger.assert_not_called()
\ No newline at end of file
+    parser3.assert_called_once_with(Path("x.tsv"))  # NOSONAR python:S1192
```
Comment (bio-boris) [inline tests/import_specifications/test_file_parser.py:538]: I think we can ignore sonarcloud, not a big worry

File: tests/import_specifications/test_file_writers.py [hunk 24]
```diff
@@ -193,84 +191,140 @@ def test_file_writers_fail():
...
+        },
+        E(
+            "Data type ty data row 1's value for parameter whee "
+            + "is not a number or a string"
+        ),
+    )
 
 
-def file_writers_fail(path: Path, types: dict, expected: Exception):
+def file_writers_fail(path: Union[Path, None], types: dict, expected: Exception):
```
Comment (bio-boris) [inline tests/import_specifications/test_file_writers.py:327]: LGTM

File: tests/pyproject.toml [hunk 25]
```diff
@@ -3,4 +3,5 @@ minversion = "6.0"
 addopts = "-ra -q"
 testpaths = [
     "tests",
-]
\ No newline at end of file
+]
+asyncio_mode = "auto"
```
Comment (bio-boris) [inline tests/pyproject.toml:7]: This is some sort of test fixture

File: tests/import_specifications/test_individual_parsers.py [hunk 26]
```diff
@@ -244,22 +368,34 @@ def test_xsv_parse_fail_empty_file(temp_dir: Path):
...
+        temp_dir, ["Data type: foo; Columns: 22; Version: 87"], parse_csv, err
+    )
 
 
 def test_xsv_parse_fail_missing_column_headers(temp_dir: Path):
     err = "Missing 2nd header line"
-    _xsv_parse_fail(temp_dir, ["Data type: foo; Columns: 3; Version: 1\n"], parse_csv, err)
+    _xsv_parse_fail(
+        temp_dir,
+        ["Data type: foo; Columns: 3; Version: 1\n"],  # NOSONAR python:S1192
```
Comment (bio-boris) [inline tests/import_specifications/test_individual_parsers.py:389]: OK

File: tests/test_app.py [hunk 27]
```diff
@@ -1578,33 +1682,39 @@ async def _write_bulk_specification_json_fail(json: Any, err: str):
...
+        {"output_directory": "foo", "output_file_type": "XSV"},
+        "Invalid output_file_type: XSV",
+    )
 
 
 async def test_write_bulk_specification_fail_invalid_type_value():
     await _write_bulk_specification_json_fail(
         {"output_directory": "foo", "output_file_type": "CSV", "types": {"a": "fake"}},
-         "The value for data type a must be a mapping"
+        "The value for data type a must be a mapping",
```
Comment (bio-boris) [inline tests/test_app.py:1717]: ok

File: tests/test_app_error_formatter.py [hunk 28]
```diff
@@ -91,21 +96,24 @@ def test_format_import_spec_errors_all_the_errors_with_tabs():
...
+            "tab": "tab1",
+        },
+        {
+            "type": "incorrect_column_count",
+            "message": "foobar2",
+            "file": "f2",
+            "tab": "tab2",
+        },
+        {
+            "type": "multiple_specifications_for_data_type",
```
Comment (bio-boris) [inline tests/test_app_error_formatter.py:112]: ok

File: tests/test_autodetect.py [hunk 29]
```diff
@@ -176,49 +214,59 @@ def test_zip_mappings():
...
-             "file_ext_type": ['CompressedFileFormatArchive']
-             },
-            {"prefix": "some.dots", "suffix": "gff3.gz", "file_ext_type": ['GFF']},
-        ]
+            {
+                "prefix": "file.name",
+                "suffix": "Gz",
+                "file_ext_type": ["CompressedFileFormatArchive"],
+            },
+            {"prefix": "some.dots", "suffix": "gff3.gz", "file_ext_type": ["GFF"]},
```
Comment (bio-boris) [inline tests/test_autodetect.py:270]: lgtm

File: tests/test_metadata.py [hunk 30]
```diff
@@ -73,8 +66,9 @@ async def _incomplete_metadata_file_update(temp_dir, metadict, source):
         "lineCount": "5",
         "name": "myfilename",
         "path": "user_path",
-        "size": 1280
-        }
+        "size": 1280,
```
Comment (bio-boris) [inline tests/test_metadata.py:69]: ok

File: tests/test_utils.py [hunk 31]
```diff
@@ -34,21 +34,23 @@ def bootstrap_config():
...
 
 def check_file_contents(file: Path, lines: list[str]):
-    with open(file) as f:
+    with open(file, "r", encoding="utf-8") as f:
         assert f.readlines() == lines
 
+
 def check_excel_contents(
-    wb: openpyxl.Workbook,
+    workbook: openpyxl.Workbook,
```
Comment (bio-boris) [inline tests/test_utils.py:46]: ok

**Review by bio-boris — APPROVED** [review]
LGTM

**Pull request merged**